src/source/shift.rs

```diff
@@ -4,141 +4,136 @@ use json;
-        console.warn(result);
-        "##,
-        script);
-
-        debug!(target: "Shift", "Launching script {}", script);
-
-        let node_memory = match env::var("NODE_MAX_OLD_SPACE_SIZE") {
-            Err(_) => String::from("--max_old_space_size=2048"),
-            Ok(v) => format!("--max_old_space_size={}", v)
-        };
+struct ScriptIO {
+    input: LineWriter<ChildStdin>,
+    output: Lines<BufReader<ChildStdout>>,
+}
 
-        let mut child = Command::new(&*self.bin_path)
-            .arg(node_memory)
-            .env("NODE_PATH", "node_modules")
+pub struct Script(Mutex<ScriptIO>);
+
+impl Script {
+    pub fn try_new(node_config: &NodeConfig, name: &str) -> Result<Self, Error> {
+        let child = node_config.create_cmd(name)
             .stdin(Stdio::piped())
             .stdout(Stdio::piped())
-            .stderr(Stdio::piped())
+            .stderr(Stdio::inherit())
             .spawn()
             .map_err(Error::CouldNotLaunch)?;
 
-        if let Some(ref mut stdin) = child.stdin {
-            stdin.write(script.as_bytes())
-                .map_err(Error::ExecutionError)?;
-        }
-
-        let output = child.wait_with_output()
-            .map_err(Error::ExecutionError)?;
+        Ok(Script(Mutex::new(ScriptIO {
+            input: LineWriter::new(child.stdin.unwrap()),
+            output: BufReader::new(child.stdout.unwrap()).lines(),
+        })))
+    }
 
-        debug!(target: "Shift", "Output {:?}", output);
-        if !output.status.success() {
-            error!(target: "Shift", "Script error {}", String::from_utf8(output.stderr).unwrap());
-            return Err(Error::ReturnedError(output.status));
+    pub fn transform(&self, input: &JSON) -> Result<JSON, Error> {
+        let output = (move || {
+            let mut io = self.0.lock().unwrap();
+            input.write(&mut io.input)?;
+            writeln!(io.input)?;
+            io.output.next().unwrap()
+        })().map_err(Error::IOError)?;
+
+        let result = json::parse(&output).map_err(Error::JsonError)?;
+        if let JSON::Object(mut obj) = result {
+            if let (Some(mut value), Some(ty)) = (obj.remove("value"), obj.get("type").and_then(|ty| ty.as_str())) {
+                match ty {
+                    "Ok" => return Ok(value),
+                    "Err" => if let Some(msg) = value.take_string() {
+                        return Err(Error::ParsingError(msg));
+                    },
+                    _ => {},
+                }
+            }
         }
-
-        let result = String::from_utf8(output.stdout)
-            .map_err(Error::InvalidUTF8)?;
-        Ok(result)
+        Err(Error::JsonError(json::Error::wrong_type("Result-like JSON object")))
     }
+}
 
-    fn parse_script_json_output(&self, script: &str) -> Result<JSON, Error> {
-        let stdout = self.parse_script_output(script)?;
+/// Using a Node + Shift binary to parse an AST.
+pub struct Shift {
+    parse_str: Script,
```

Comment:
If I read this correctly, we have three processes launched whenever we call `try_new`, right?

Could you document this? Also, as an optimization, we could probably create them lazily.

Reply:
I guess we could, but for now this is in the worst case taking same time as before the PR (when one invocation == one reencoded script), and in the best case you create parser just once and process lots of scripts anyway, and then process creation time is now negligible.

Reply:
Sure. Just noticing a low-hanging fruit :)

Reply:
(don't hesitate to make this a followup issue)